src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs

```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+                return Sse42.Crc32(crc, data);
+            }
+            if (Crc32.IsSupported)
+            {
+                // uint32_t __crc32ch (uint32_t a, uint16_t b)
+                return Crc32.ComputeCrc32C(crc, data);
+            }
+
+            // Software fallback
+            return Crc32Fallback.Crc32C(crc, data);
+        }
+
+        /// <summary>
+        /// Accumulates the CRC (Cyclic redundancy check) checksum.
+        /// </summary>
+        /// <param name="crc">The base value to calculate checksum on</param>
+        /// <param name="data">The data for which to compute the checksum</param>
+        /// <returns>The CRC-checksum</returns>
+        [CLSCompliant(false)]
+        public static uint Crc32C(uint crc, uint data)
+        {
+            if (Sse42.IsSupported)
+            {
+                // unsigned int _mm_crc32_u32 (unsigned int crc, unsigned int v)
+                return Sse42.Crc32(crc, data);
+            }
+            if (Crc32.IsSupported)
+            {
+                // uint32_t __crc32ch (uint32_t a, uint32_t b)
+                return Crc32.ComputeCrc32C(crc, data);
+            }
+
+            // Software fallback
+            return Crc32Fallback.Crc32C(crc, data);
+        }
+
+        /// <summary>
+        /// Accumulates the CRC (Cyclic redundancy check) checksum.
+        /// </summary>
+        /// <param name="crc">The base value to calculate checksum on</param>
+        /// <param name="data">The data for which to compute the checksum</param>
+        /// <returns>The CRC-checksum</returns>
+        [CLSCompliant(false)]
+        public static uint Crc32C(uint crc, ulong data)
+        {
+            if (Sse42.X64.IsSupported)
+            {
+                return (uint)Sse42.X64.Crc32(crc, data);
+            }
+            if (Sse42.IsSupported)
+            {
+                return Sse42.Crc32(Sse42.Crc32(crc, (uint)data), (uint)(data >> 32));
+            }
+            if (Crc32.Arm64.IsSupported)
+            {
+                // uint32_t __crc32cd (uint32_t a, uint64_t b)
+                return Crc32.Arm64.ComputeCrc32C(crc, data);
+            }
+
+            // Software fallback
+            return Crc32Fallback.Crc32C(crc, data);
+        }
+
+        private static class Crc32Fallback
+        {
+            private static readonly uint[] s_crcTable = Crc32ReflectedTable.Generate(0x82F63B78u);
+
+            internal static uint Crc32C(uint crc, byte data)
+            {
+                ref uint lookupTable = ref MemoryMarshal.GetArrayDataReference(s_crcTable);
+                crc = Unsafe.Add(ref lookupTable, (nint)(byte)(crc ^ data)) ^ (crc >> 8);
+
+                return crc;
+            }
+
+            internal static uint Crc32C(uint crc, ushort data)
+            {
+                if (!BitConverter.IsLittleEndian)
+                {
+                    data = BinaryPrimitives.ReverseEndianness(data);
+                }
```

Comment:
Is this correct?

If I define `ushort x = 0x0201`, why should big endian process these as `0x02` then `0x01`? We have an already in register 16-bit value, so any endianness fixups should have been handled by the user already, no?

Reply:
Presumably, the caller would be reading the value from memory (in host order), not setting it as a literal.  But it's a good question, since I assume arm64 running in BE mode would read from memory in BE order, but then the crc32 intrinsic assumes LE within the register.  I suppose if the byte order fixup is handled here, it should be done in the outer method and not just in the fallback.

The tests added with this are using literals as well, so I'd assume those would fail on the s390 pipelines as is.  Perhaps a more realistic span-based test would be in order, whether the endianness issue is handled by these methods or by the caller.

Reply:
> Presumably, the caller would be reading the value from memory (in host order), not setting it as a literal. But it's a good question, since I assume arm64 running in BE mode would read from memory in BE order, but then the crc32 intrinsic assumes LE within the register. I suppose if the byte order fixup is handled here, it should be done in the outer method and not just in the fallback.

The general point is we're taking an input "as is", so it should be presumed to already be in the correct order with any fixups done by the caller. As it is I believe BigEndian will return a different result for the same input as compared to LittleEndian